src/sinks/socket.rs

```diff
@@ -155,4 +155,106 @@ mod test {
+    //
+    // This test basically sends 10 events shutsdown the server and forces a
+    // reconnect. It then forces another 10 events through and we should get a
+    // total of 20 events.
+    //
+    // If this test hangs that means somewhere we are not collecting the correct
+    // events.
+    #[test]
+    fn tcp_stream_detects_disconnect() {
+        let addr = next_addr();
+        let config = SocketSinkConfig {
+            mode: Mode::Tcp(TcpSinkConfig {
+                address: addr.to_string(),
+                encoding: Encoding::Text.into(),
+                // TODO: enable TLS here since this is where
+                // ran into the issue before!
+                tls: None,
+            }),
+        };
+        let mut rt = runtime();
+        let context = SinkContext::new_test(rt.executor());
+        let (sink, _healthcheck) = config.build(context).unwrap();
+
+        use std::sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc,
+        };
+
+        let close = Arc::new(tokio::sync::Notify::new());
+        let counter = Arc::new(AtomicUsize::new(0));
+
+        let close1 = close.clone();
+        let counter1 = counter.clone();
+        let jh = rt.spawn_handle(async move {
+            use tokio::{io::AsyncReadExt, net::TcpListener};
+
+            let mut listener = TcpListener::bind(&addr).await.unwrap();
+
+            // Only accept two connections after the second connection is done
+            // we can exit.
+            for _ in 0..2 {
+                let (mut conn, _) = listener.accept().await.unwrap();
+
+                loop {
+                    let mut buf = vec![0u8; 11];
+
+                    tokio::select! {
+                        _ = close1.notified() => {
+                            conn.shutdown(std::net::Shutdown::Write).unwrap();
+                            break;
+                        }
+
+                        res = conn.read(&mut buf) => {
+                            let n = res.unwrap();
+
+                           if  n == 0 {
+                               break;
+                           } else {
+                               counter1.fetch_add(1, Ordering::Relaxed);
+                           }
+                        }
+                    };
+                }
+            }
+        });
+
+        let (_, events) = random_lines_with_stream(10, 10);
+        let pump = sink.send_all(events);
+        let (sink, _) = rt.block_on(pump).unwrap();
+
+        // Loop and check for 10 events, we should always get 10 events. Once,
+        // we have 10 events we can tell the server to shutdown to simulate the
+        // remote shutting down on an idle connection.
+        for _ in 0..100 {
+            let amnt = counter.load(Ordering::SeqCst);
+
+            if amnt == 10 {
+                close.notify();
+                break;
+            }
+
+            std::thread::sleep(std::time::Duration::from_millis(100));
```

Comment:
I like to think we could do this without sleeping. Maybe very shortly at the end before checking the final count?

Reply:
We need to sleep to ensure the server read all 10 events. This kinda simulates an idle timeout. So we want to ensure the 10 got there before shutdown then shutdown then send another 10 which should trigger the reconnect process. Then check another 10.